File: src/session/undo.ts
import { mkdirSync, readFileSync, readdirSync, rmSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { ensureDir } from '../config/config.ts';

interface Snapshot {
  file: string;
  previousContent: string | null; // null = file did not exist before
  at: string;
}

/**
 * Records the pre-modification content of files the agent changes, so the
 * user can `/undo` the most recent change. Backed by `.myagent/undo/`.
 */
export class UndoStore {
  private readonly dir: string;
  private counter = 0;

  constructor(projectRoot: string) {
    this.dir = join(projectRoot, '.myagent', 'undo');
    ensureDir(this.dir);
    this.counter = readdirSync(this.dir).filter((f) => f.endsWith('.json')).length;
  }

  /** Record a file before it is modified. Returns true when a snapshot was stored. */
  record(file: string, previousContent: string | null): boolean {
    this.counter += 1;
    const stamp = new Date().toISOString().replace(/[:.]/g, '-');
    const path = join(this.dir, `${String(this.counter).padStart(6, '0')}_${stamp}.json`);
    const snap: Snapshot = { file, previousContent, at: new Date().toISOString() };
    try {
      writeFileSync(path, JSON.stringify(snap, null, 2), 'utf8');
      return true;
    } catch {
      return false;
    }
  }

  /** Revert the most recent snapshot. Returns a description, or null if empty. */
  undo(): { file: string; hadPrevious: boolean } | null {
    const snaps = readdirSync(this.dir)
      .filter((f) => f.endsWith('.json'))
      .sort();
    if (snaps.length === 0) return null;
    const latest = snaps[snaps.length - 1] as string;
    const path = join(this.dir, latest);
    let snap: Snapshot;
    try {
      snap = JSON.parse(readFileSync(path, 'utf8')) as Snapshot;
    } catch {
      return null;
    }
    if (snap.previousContent === null) {
      // File was newly created by the agent — remove it.
      rmSync(snap.file, { force: true });
    } else {
      mkdirSync(dirname(snap.file), { recursive: true });
      writeFileSync(snap.file, snap.previousContent, 'utf8');
    }
    rmSync(path, { force: true });
    return { file: snap.file, hadPrevious: snap.previousContent !== null };
  }

  /** All pending snapshot descriptions (used by /diff when there is no git repo). */
  pendingChanges(): string[] {
    return readdirSync(this.dir)
      .filter((f) => f.endsWith('.json'))
      .sort()
      .map((f) => {
        try {
          const snap = JSON.parse(readFileSync(join(this.dir, f), 'utf8')) as Snapshot;
          return snap.previousContent === null ? `+ ${snap.file} (created)` : `~ ${snap.file} (modified)`;
        } catch {
          return '';
        }
      })
      .filter((s) => s.length > 0);
  }

  get count(): number {
    return readdirSync(this.dir).filter((f) => f.endsWith('.json')).length;
  }
}

File: grace/session/undo.py
"""Undo snapshots (port of src/session/undo.ts)."""

import json
import os
from pathlib import Path


class UndoStore:
    """Records the pre-modification content of files the agent changes, so the
    user can `/undo` the most recent change. Backed by `.zeesh/undo/`."""

    def __init__(self, project_root: str) -> None:
        self.dir = os.path.join(project_root, ".zeesh", "undo")
        Path(self.dir).mkdir(parents=True, exist_ok=True)
        self.counter = len([f for f in os.listdir(self.dir) if f.endswith(".json")])

    def record(self, file: str, previous_content: str | None) -> bool:
        self.counter += 1
        stamp = __import__("datetime").datetime.now().isoformat().replace(":", "-").replace(".", "-")
        path = os.path.join(self.dir, f"{self.counter:06d}_{stamp}.json")
        snap = {"file": file, "previousContent": previous_content, "at": __import__("datetime").datetime.now().isoformat()}
        try:
            Path(path).write_text(json.dumps(snap, indent=2), encoding="utf-8")
            return True
        except Exception:
            return False

    def undo(self) -> dict | None:
        snaps = sorted(f for f in os.listdir(self.dir) if f.endswith(".json"))
        if not snaps:
            return None
        latest = snaps[-1]
        path = os.path.join(self.dir, latest)
        try:
            snap = json.loads(Path(path).read_text(encoding="utf-8"))
        except Exception:
            return None
        if snap.get("previousContent") is None:
            # File was newly created by the agent — remove it.
            try:
                os.remove(snap["file"])
            except OSError:
                pass
        else:
            Path(os.path.dirname(snap["file"])).mkdir(parents=True, exist_ok=True)
            Path(snap["file"]).write_text(snap["previousContent"], encoding="utf-8")
        try:
            os.remove(path)
        except OSError:
            pass
        return {"file": snap["file"], "hadPrevious": snap.get("previousContent") is not None}

    def pending_changes(self) -> list[str]:
        out = []
        for f in sorted(x for x in os.listdir(self.dir) if x.endswith(".json")):
            try:
                snap = json.loads(Path(os.path.join(self.dir, f)).read_text(encoding="utf-8"))
            except Exception:
                continue
            if snap.get("previousContent") is None:
                out.append(f"+ {snap['file']} (created)")
            else:
                out.append(f"~ {snap['file']} (modified)")
        return out

    @property
    def count(self) -> int:
        return len([f for f in os.listdir(self.dir) if f.endswith(".json")])
